src/inmanta/server/services/orchestrationservice.py

```diff
@@ -419,6 +426,138 @@ def safe_get(input: JsonType, key: str, default: object) -> object:
+                    if key == Id.parse_id(res["id"]).resource_str():
+                        pair = (partial_update, res, resource_set, old_resource[1])
+                paired_resources.append(pair)
+
+            return paired_resources
+
+        def get_updated_resource_sets(paired_resources: List[tuple[[Dict[str, Any], [Dict[str, Any]], str, str]]]) -> Set[str]:
+            result: Set[str] = set()
+            for res in paired_resources:
+                resource_set = res[2]
+                if resource_set:
+                    result.add(resource_set)
+            return result
+
+        async def get_old_resources():
+            old_data = await data.Resource.get_list()
+            result: List[tuple[[Dict[str, Any]], str]] = []
+            for res in old_data:
+                resource = {
+                    "id": res.resource_version_id,
+                }
+                resource.update(res.attributes)
+                result.append((resource, res.resource_set))
+            return result
+
+        old_resources = await get_old_resources()
+        paired_resources = pair_resources_partial_update_to_old_version(old_resources, partial_updates)
+        updated_resource_sets = get_updated_resource_sets(paired_resources)
+
+        def copy_with_incremented_version(resource):
+            res = Id.parse_id(resource["id"])
+            res.increment_version()
+            resource["id"] = res.resource_version_str()
+            return resource
+
+        to_keep: List[Dict[str, Any]] = [
+            copy_with_incremented_version(r[0])
+            for r in old_resources
+            if not r[1] in removed_resource_sets and (r[1] is None or not r[1] in updated_resource_sets)
+        ]
+
+        result: Dict[ResourceIdStr, Dict[str, Any]] = {r["id"]: r for r in to_keep}
+
+        for resource_partial, resource_old, resource_set_partial, resource_set_old in paired_resources:
+            assert resource_partial is not None
+            assert (
+                resource_old is None
+                or Id.parse_id(resource_old["id"]).resource_str() == Id.parse_id(resource_partial["id"]).resource_str()
+            )
+            if resource_old is None:
+                result[resource_partial["id"]] = resource_partial
+            else:
+                if resource_set_partial != resource_set_old:
+                    raise BadRequest(
+                        f"A partial compile cannot migrate a resource({resource_partial['id']}) to another resource set"
+                    )
+                if resource_set_partial == None and resource_partial != resource_old:  # todo:: is the != comparison enough?
+                    raise BadRequest(
+                        f"Resource ({resource_partial['id']}) without a resource set cannot be updated via a partial compile"
+                    )
+                else:
+                    result[resource_partial["id"]] = resource_partial
+        return list(result.values())
```

Comment:
I think I could just use an array from the start. Should I change this?

Reply:
Or you could just change the return type to `collections.abc.Sequence`, I believe `dict.values()` is compatible with that.

Reply:
Oh no, `.values()` appears to be some sort of generator:
```python
>>> dct = {"a": 1, "b": 2}
>>> l = dct.values()
>>> l
dict_values([1, 2])
>>> del dct["a"]
>>> l
dict_values([2])
```